### PR TITLE
settings: combine "Do not show" flags into one setting

### PR DESCRIPTION
**Pull request by gsweetwood**
## Problem
A setting for each prompt that can be disabled will be too noisy.
## Solution
Condense the prompts that can be disabled into a single setting and provide utilities to read and disable the setting.

# Screenshot 
UX
multi checkbox setting demo

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/utilities/promptSetting.ts
Comment (justinmk3) [inline src/shared/utilities/promptSetting.ts:1]: the codebase already has `src/shared/settingsConfiguration.ts`, what is the reason for adding this separate module ?

File: package.json
Comment (justinmk3) [inline package.json:190]: if we explicitly name the keys in the default here, that is a maintenance cost: we must always remember to update this. Can we not instead treat a missing key as `true`, in the code logic?

File: package.json
Comment (justinmk3) [inline package.json:186]: other localization ids in this same file, use the `AWS.configuration.*` namespace. What is the reason for choosing `AWS.setting` (not used anywhere in the codebase)?
```suggestion
                    "description": "%AWS.configuration.togglePrompts.description%",
```

File: src/shared/utilities/promptSetting.ts
Comment (justinmk3) [inline src/shared/utilities/promptSetting.ts:9]: types are not named with all caps. You can check this via `git grep! -E 'type [A-Z][A-Z]+'`.

PROMPT_NAME is a questionable name here, because (1) we have an unrelated "Prompt" concept elsewhere in the codebase, and (2) this is a setting. This raises the question (from many previous PRs) of why we are naming this at all? It could be defined inline in the `disablePrompt()` signature.

File: src/shared/utilities/promptSetting.ts
Comment (justinmk3) [inline src/shared/utilities/promptSetting.ts:16]: This is not needed unless we plan to introduce a prompt that defaults to `false`--which would be very unlikely, and is something we should wait for rather than writing code to anticipate it.

File: src/shared/utilities/promptSetting.ts
Comment (justinmk3) [inline src/shared/utilities/promptSetting.ts:27]: - vscode settings cannot be trusted to return the type given by the type parameter. The user's settings.json could have any JSON value (boolean, string, array, ...). Rather than try/catch we can test for this to show a more informative message.
    - Because we own this setting (it's an "AWS" setting), we can just overwrite a bad value.
- writeSetting() already does try/catch internally, so the try/catch in this scope is unnecessary, if we check the type.

```suggestion
        if (type prompts !== 'object' || Array.isArray(prompts)) {
            getLogger().warn('setting "togglePrompts" has an unexpected type. Overwriting.')
            prompts = {}
        }
        prompts[promptName] = false
```

File: src/shared/utilities/promptSetting.ts
Comment (justinmk3) [inline src/shared/utilities/promptSetting.ts:28]: we can avoid unnecessary updates to the user's settings file: before doing `prompts[promptName] = false`, check if we are actually changing anything.

```
if (!prompts.includes(promptName) || prompts[promptName] !== false) {
    settings.writeSetting(...)
}
```

File: src/shared/utilities/promptSetting.ts
Comment (justinmk3) [inline src/shared/utilities/promptSetting.ts:23]: This function could be generalized to "disable" or "enable" at very little cost, so generally it's worth doing so.

File: src/shared/utilities/promptSetting.ts
Comment (justinmk3) [inline src/shared/utilities/promptSetting.ts:25]: `readPromptSetting()` was defined, but it's not being used here?

File: src/shared/utilities/promptSetting.ts
Comment (justinmk3) [inline src/shared/utilities/promptSetting.ts:35]: vscode settings cannot be trusted to return the type given by the type parameter. The user's settings.json could have any JSON value (boolean, string, array, ...). Rather than try/catch we can test for this to show a precise message.

```
if (type promptSetting !== 'object' || Array.isArray(promptSetting)) {
    getLogger().warn('setting "togglePrompts" has an unexpected type. Falling back to default.')
    return true
}
```

**Review by justinmk3 — CHANGES_REQUESTED** [review]

File: src/shared/utilities/promptSetting.ts
Comment (JadenSimon) [reply src/shared/utilities/promptSetting.ts:20]: This interface is a narrowing of the TS object type since it restricts any property to be a boolean, so it isn't entirely pointless. This interface can potentially be much more useful when TS 4.4 is released, allowing the index signature to be further narrowed with the constant prompt name alias.

This type does not need to be exported though since it just serves to keep things internally consistent. Same goes for `PROMPT_NAME` unless there is a separate module that is manipulating arbitrary prompt names? Callers of `disablePrompt` do not need to know how the data is being represented within the settings.

Comment (JadenSimon) [reply src/shared/utilities/promptSetting.ts:20]: After looking over the code again I see that the type is only being used to read from settings, so yes I would say that just assuming the settings has the correct shape is poor practice.

File: src/shared/utilities/promptSetting.ts
Comment (JadenSimon) [reply src/shared/utilities/promptSetting.ts:9]: I'm guessing the rationale behind this was to have any callers add their own 'key' to the union type. But that seems rather cumbersome IMO. Consumers of the module should not need to change it to be able to use it. I suppose it'd be better to just use a similar style to `localize` where callers need to define their own key?

File: src/shared/utilities/promptSetting.ts
Comment (JadenSimon) [reply src/shared/utilities/promptSetting.ts:27]: I wonder if the config contribution point can accept a schema to validate individual settings? We should still validate within the code too but both would be ideal.

File: package.nls.json
Comment (bryceitoc9) [inline package.nls.json:593]: Toggle implies that you can turn these messages off. I'd suggest something like "Restore all \"Do not show again\" notifications"

File: package.json
Comment (bryceitoc9) [inline package.json:185]: What if this were an array? This would remove a fair amount of the concerns regarding true/false/non-boolean typing. Might be a bit more understandable to the user, especially if this was renamed `aws.doNotShowPrompts` or something.

**Review by bryceitoc9 — COMMENTED** [review]
My big feedback here is why not just make this an array which contains prompts to not show again. I'm thinking this would be more accessible to the user, and would simplify a lot of steps here.

File: package.json
Comment (gsweetwood) [reply package.json:185]: I think you're suggestion of an array is great.  Originally I was thinking of mapping each prompt, but an array of disabled prompts is cleaner and is easier to append.  It eliminates any maintenance cost for future disable prompts.

File: src/shared/utilities/promptSetting.ts
Comment (justinmk3) [reply src/shared/utilities/promptSetting.ts]: I think it makes sense to have the valid keys defined in one place.

File: package.json
Comment (gsweetwood) [reply package.json:186]: corrected

File: src/shared/utilities/promptSetting.ts
Comment (gsweetwood) [reply src/shared/utilities/promptSetting.ts]: All new types and interfaces removed, not needed here.

File: package.json
Comment (JadenSimon) [inline package.json:185]: Since this is an array of strings now, I believe we can define the type as such:
```json
"type": ["string"],
```

File: src/shared/utilities/promptSetting.ts
Comment (gsweetwood) [reply src/shared/utilities/promptSetting.ts]: This function is now a method of DefaultSettingsConfiguration and is called `disable`

File: src/shared/utilities/promptSetting.ts
Comment (gsweetwood) [reply src/shared/utilities/promptSetting.ts]: This was very useful, thank you.

File: package.nls.json
Comment (gsweetwood) [reply package.nls.json]: Changed to "List of prompts not to show again"  This can be reset in the settings page if the user hovers over the setting where a little gear pops up which is a menu.  They can 'Reset setting' which  deletes the array from the json file which works fine. The setting will be added anytime the `disable` method is called.

File: package.json
Comment (gsweetwood) [reply package.json]: Changed to "array".  Your suggestion doesn't seem to work correctly.  I think we can leave off the type.

Comment (JadenSimon) [reply package.json]: Hmm not sure then. The docs have "type": ["string", "null"] as an example so figured it would work.

File: package.nls.json
Comment (kiiadi) [inline package.nls.json:40]: wording suggestion: "List of prompts to suppress"

File: src/shared/settingsConfiguration.ts
Comment (kiiadi) [inline src/shared/settingsConfiguration.ts:83]: where do these actually get called?

**Review by kiiadi — COMMENTED** [review]
How does this work in cases where the suppression was off the back of a choice (and the suppression is to effectively remember the choice)?

It's a pretty typical pattern: "Answer this question, and if you want us to use the same answer in future, tick this box"

File: src/shared/settingsConfiguration.ts
Comment (gsweetwood) [reply src/shared/settingsConfiguration.ts:83]: Once this is merged, they will begin to be used in #1908,.

File: package.nls.json
Comment (gsweetwood) [reply package.nls.json:40]: Thanks for the suggestion. Definitely an improvement.  Change added.

**Comment by JadenSimon** [comment]
I was wrong about not being able to do 'checkboxes' for settings. Apparently it's a pretty new feature (and undocumented):
https://code.visualstudio.com/updates/v1_58#_settings-editor

**Comment by gsweetwood** [comment]
@kiiadi Prompts that require a choice are an interruption in a process to potentially warn/notify a user and asks if they would like to continue.  Clicking yes proceeds, no exits.  If they click 'Yes, and don't ask again", it will proceed and any subsequent calls will skip the prompt and proceed.

**Comment by gsweetwood** [comment]
Pushed a refactor to use the new multi checkbox settings option.  Next commit will adjust the tests.

File: src/shared/settingsConfiguration.ts
Comment (justinmk3) [reply src/shared/settingsConfiguration.ts:76]: it probably makes sense to also convert existing suppress-prompt calls in this PR, there are not too many cases. For example 

https://github.com/aws/aws-toolkit-vscode/blob/e8af78cc02735411718ef982f364c1df3c473228/src/apprunner/wizards/deploymentButton.ts#L37

and

https://github.com/aws/aws-toolkit-vscode/blob/e8af78cc02735411718ef982f364c1df3c473228/src/shared/sam/activation.ts#L346

File: src/shared/settingsConfiguration.ts
Comment (justinmk3) [inline src/shared/settingsConfiguration.ts:85]: fully-qualified name is less indirection at little cost.
```suggestion
     * Gets the 'aws.suppressPrompts' setting. This will reset the setting to default if it does not
```

File: src/test/shared/settingsConfiguration.test.ts
Comment (justinmk3) [inline src/test/shared/settingsConfiguration.test.ts:80]: the comment is more confusing than useful, IMO. Or I don't understand it. Tests for different functions don't need to explain that they're testing different functions.

File: src/test/shared/settingsConfiguration.test.ts
Comment (justinmk3) [inline src/test/shared/settingsConfiguration.test.ts:87]: why use sinon? the pattern shown on lines 20-30 should work for this setting also. Just use a different SETTINGS_KEY.

File: src/shared/settingsConfiguration.ts
Comment (justinmk3) [inline src/shared/settingsConfiguration.ts:63]: - disable() is too vague of a name, it lives on a generic Settings class. It is suppressing prompts.
- as suggested in https://github.com/aws/aws-toolkit-vscode/pull/1937#discussion_r687092120 , this could be parameterized at near-zero cost, so we should do it to avoid potential refactor in the future.

```suggestion
    public enablePrompt(promptName: string, enable: boolean): void {
```

File: src/shared/settingsConfiguration.ts
Comment (justinmk3) [inline src/shared/settingsConfiguration.ts:93]: condition is too loose. if `setting` is undefined we don't need to enter this if-block and rewrite the settings. That would result in forcibly adding the setting even for users who don't disable the prompts. It's also a small perforfmance cost.
```suggestion
            if (setting !== undefined && (typeof setting !== 'object' || typeof setting[promptName] !== 'boolean')) {
```

File: src/test/shared/settingsConfiguration.test.ts
Comment (gsweetwood) [reply src/test/shared/settingsConfiguration.test.ts:80]: The tests above that point were reading and writing to the actual settings file.  This was to point out these tests needed to mock the read and write calls, but it's not necessary to mention so it can be removed.

File: src/shared/settingsConfiguration.ts
Comment (gsweetwood) [reply src/shared/settingsConfiguration.ts:63]: That's what it was called previously when it was in a separate module and you made this comment [here](https://github.com/aws/aws-toolkit-vscode/pull/1937#discussion_r687092120) but now that it's in this class it makes sense to call it something more specific.

File: src/shared/settingsConfiguration.ts
Comment (gsweetwood) [reply src/shared/settingsConfiguration.ts:93]: This is a good point and the rest of the flow already allows for undefined values, so this change makes sense.

File: src/shared/settingsConfiguration.ts
Comment (justinmk3) [reply src/shared/settingsConfiguration.ts]: > you made this comment [here](https://github.com/aws/aws-toolkit-vscode/pull/1937#discussion_r687092120)

by 'generalized to "disable" or "enable"' I meant the function could be generalized to perform both operations. Instead of only performing disable. I didn't mean to change the name.

File: src/shared/settingsConfiguration.ts
Comment (justinmk3) [reply src/shared/settingsConfiguration.ts]: > UPDATE: I don't think `setting` can ever be undefined.

Then we should change the return type of `readSetting()` (separate PR).

File: src/shared/settingsConfiguration.ts
Comment (gsweetwood) [reply src/shared/settingsConfiguration.ts]: Thanks for the clarification.

File: src/test/shared/settingsConfiguration.test.ts
Comment (JadenSimon) [reply src/test/shared/settingsConfiguration.test.ts:87]: Friendly reminder to everyone that refactoring tests to reduce copy-paste code _is_ important for maintainability. Your test code should not be an afterthought.

File: src/shared/settingsConfiguration.ts
Comment (justinmk3) [inline src/shared/settingsConfiguration.ts:97]: We can just return `false` here. That simplifies the return type and simplifies caller logic. 

`undefined` could be useful if callers were interested in whether the setting existed at all. But I don't think there are such cases.

Comment (gsweetwood) [reply src/shared/settingsConfiguration.ts:97]: `False` will not work here. The purpose of this function is to return the setting object, which has all the prompts. Unlike our other settings, we can't just overwrite it.  We need to get the object, update the one prompt, then write that back to settings. We should use the `shouldDisplayPrompt` function when we need to determine if a prompt is suppressed.

Comment (justinmk3) [reply src/shared/settingsConfiguration.ts:97]: oops i misread the return type.

File: src/shared/settingsConfiguration.ts
Comment (gsweetwood) [reply src/shared/settingsConfiguration.ts]: These have been added to this PR.

File: src/shared/sam/activation.ts
Comment (justinmk3) [inline src/shared/sam/activation.ts:253]: The call site was not updated to `await` this. This function checks `vscode.window.visibleTextEditors` so there might be a race. When changing functions to `async`, always check their call sites.

File: package.json
Comment (gsweetwood) [reply package.json:189]: If we want to keep the multi select checkbox list, then no, I believe they need to be listed here. This gives the customer greater visibility and control, but does add to maintenance costs. The previous version of this PR was the opposite, no maintenance cost but the customer had to either reset all prompts or edit the settings.json manually.

Comment (justinmk3) [reply package.json:191]: Doesn't seem like the right tradeof because almost no one will ever change these settings. But I don't want to block this longer, so let's merge it.

**Pull request merged**
